### PR TITLE
Only compress responses if request was compressed

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is a follow-up to some discussions around #36399. Currently we have
relatively confusing compression behavior where compression can be
configured for requests based on `transport.compress` or a specific
setting for a remote cluster. However, we can only compress responses
based on `transport.compress` as we do not know where a request is
coming from (currently).

This commit modifies the behavior to NEVER compress responses based on
settings. Instead, a response will only be compressed if the request was
compressed. This commit also updates the documentation to more clearly
described transport level compression.

### CODE REVIEW

**Comment by tbrooks8** [comment]
This PR is based on a conversation with I had with @jasontedor where I suggested it might be nicer to have simpler rules for compressing responses. Instead of having this odd edge-case where only some of the compression settings apply to responses, none of the compression settings will apply to responses 🤣.

I wanted to open this PR and see if others agreed. If we all agree and generally there is agreement that the description about compression I added to the documentation is correct, I'll add @lcawl as a specific reviewer to iron-out the language. But @lcawl - you can hold off reviewing until we agree that this is a change we want to make.

**Comment by ywelsch** [comment]
The change makes sense to me. Can you also add tests that check that we're getting the desired behavior with this change?

File: docs/reference/modules/transport.asciidoc [hunk 1]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
+
+By default Elasticsearch network-level compression is disabled. This default
+normally makes sense for local cluster communication as compression has a
+noticeable CPU cost and local clusters tend to be setup with fast network
+connections between nodes. If compression is necessary for remote cluster
+connections only, this can be configured on a per-cluster basis using the
+`cluster.remote.${cluster_alias}.transport.compress` setting.
+
+When compression is enabled using either `transport.compress` or for a specific
+remote cluster, only the outbound requests are compressed. Elasticsearch only
```
Comment (jasontedor) [inline docs/reference/modules/transport.asciidoc]: `Elasticsearch only` -> `Elasticsearch`

**Review by jasontedor — APPROVED** [review]
LGTM.

**Review by original-brownbear — APPROVED** [review]
LGTM2

**Comment by tbrooks8** [comment]
@lcawl - I think that we're at a point where it is probably worth you reviewing the documentation.

**Review by s1monw — APPROVED** [review]

File: docs/reference/modules/transport.asciidoc [hunk 2]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
 
+[float]
+==== Transport Compression
+
+By default Elasticsearch network-level compression is disabled. This default
+normally makes sense for local cluster communication as compression has a
+noticeable CPU cost and local clusters tend to be setup with fast network
+connections between nodes. If compression is necessary for remote cluster
+connections only, this can be configured on a per-cluster basis using the
+`cluster.remote.${cluster_alias}.transport.compress` setting.
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
<<remote-cluster-settings,`cluster.remote.${cluster_alias}.transport.compress` setting>>.
```

File: docs/reference/modules/transport.asciidoc [hunk 3]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
 cause idle connections to be closed, or by setting `transport.ping_schedule` if
 keepalives cannot be configured.
 
+[float]
+==== Transport Compression
+
+By default Elasticsearch network-level compression is disabled. This default
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
By default, network-level compression is disabled in Elasticsearch. This default
```

File: docs/reference/modules/transport.asciidoc [hunk 4]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
+normally makes sense for local cluster communication as compression has a
+noticeable CPU cost and local clusters tend to be setup with fast network
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
noticeable CPU cost and local clusters tend to be set up with fast network
```

File: docs/reference/modules/transport.asciidoc [hunk 5]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
+connections between nodes. If compression is necessary for remote cluster
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
connections between nodes. If you want to compress remote cluster
```

File: docs/reference/modules/transport.asciidoc [hunk 6]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
+connections only, this can be configured on a per-cluster basis using the
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]: The "only" here seems to imply that setting transport.compress affects both internode and remote cluster communications. i.e. you set this cluster-level setting if you *only* want to compress remote cluster connections.  If that's not right, we should remove the "only".

File: docs/reference/modules/transport.asciidoc [hunk 7]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
+`cluster.remote.${cluster_alias}.transport.compress` setting.
+
+When compression is enabled using either `transport.compress` or for a specific
+remote cluster, only the outbound requests are compressed. Elasticsearch compresses
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
remote cluster, only the outbound requests are compressed. 
```

File: docs/reference/modules/transport.asciidoc [hunk 8]
```diff
@@ -102,6 +102,20 @@ and ensuring that the keepalive interval is shorter than any timeout that might
...
+responses if and only-if the inbound request received was compressed.
```
Comment (lcawl) [inline docs/reference/modules/transport.asciidoc]:
```suggestion
If compression is enabled and an inbound request is compressed, {es} also compresses the response.
```

Comment (lcawl) [reply docs/reference/modules/transport.asciidoc]: Not sure if it's true that the compression of responses requires compression to be enabled, but it seemed like a reasonable assumption.

**Comment by tbrooks8** [comment]
@lcawl - I made some changes. I went ahead and split it into request and response sections to be completely clear.

**Review by lcawl — APPROVED** [review]
LGTM

**Comment by tbrooks8** [comment]
Thanks @lcawl

**Pull request merged**